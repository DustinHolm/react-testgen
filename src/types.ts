export interface ImportBlock {
    sourceFile: string,
    isInternal: boolean,
    imports: Import[]
}

export interface Import {
    isDefault: boolean,
    name: string,
    element?: ExportableElement
}

export interface Export {
    name: string,
    isDefault: boolean,
    element?: ExportableElement
}

export enum ExportType {
    Class,
    Function,
    Interface,
    Variable
}

export interface ExportableElement {
    type: ExportType,
    returnsJSX: boolean
}

export interface ClassElement extends ExportableElement {
    methods: MethodElement[],
    constructors: Constructor[]
}

export interface MethodElement extends FunctionElement {
    name: string
}

export interface FunctionElement extends ExportableElement {
    returnType: string,
    parameters: Attribute[]
}

export interface InterfaceElement extends ExportableElement {
    attributes: Attribute[]
}

export interface VariableElement extends ExportableElement {
    varType: string,
    varValue: string
}

export interface Attribute {
    name: string,
    type: string
}

export interface Constructor {
    parameters: Attribute[]
}