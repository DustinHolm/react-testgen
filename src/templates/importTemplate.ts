import { ImportBlock } from "../types"

export const createImport = (importBlock: ImportBlock): string => {
    const defaultName = importBlock.imports.find(i => i.isDefault)?.name ?? ""
    const namedImports = importBlock.imports.filter(i => !i.isDefault)
        .map(i => i.name)
        .sort()
    const importsObject = namedImports.length > 0
        ? `{ ${namedImports.join(", ")} }`
        : ""
    const separator = defaultName !== "" && importsObject !== "" ? ", " : ""
    const fileName = importBlock.sourceFile

    return `import ${defaultName}${separator}${importsObject} from "${fileName}"`
}

export const createDefaultImports = (): string => {
    return 'import { render, screen } from "@testing-library/react"'
}