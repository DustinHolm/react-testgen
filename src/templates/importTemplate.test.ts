import { ImportBlock } from "../types"
import { createImport } from "./importTemplate"

describe("importTemplate Tests", () => {
    let givenBlock: ImportBlock

    beforeEach(() => {
        givenBlock = {
            sourceFile: "./test",
            isInternal: true,
            imports: []
        }
    })

    test("With only default import", () => {
        givenBlock.imports.push({ name: "TestDefault", isDefault: true })
        const returnedString = createImport(givenBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With only one named import", () => {
        givenBlock.imports.push({ name: "TestNamed", isDefault: false })
        const returnedString = createImport(givenBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With only multiple named imports", () => {
        givenBlock.imports.push({ name: "TestNamed_1", isDefault: false })
        givenBlock.imports.push({ name: "TestNamed_2", isDefault: false })
        const returnedString = createImport(givenBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With only multiple unsorted named imports", () => {
        givenBlock.imports.push({ name: "TestNamed_2", isDefault: false })
        givenBlock.imports.push({ name: "TestNamed_1", isDefault: false })
        const returnedString = createImport(givenBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With mixed imports", () => {
        givenBlock.imports.push({ name: "TestDefault", isDefault: true })
        givenBlock.imports.push({ name: "TestNamed_1", isDefault: false })
        givenBlock.imports.push({ name: "TestNamed_2", isDefault: false })
        const returnedString = createImport(givenBlock)

        expect(returnedString).toMatchSnapshot()
    })
})