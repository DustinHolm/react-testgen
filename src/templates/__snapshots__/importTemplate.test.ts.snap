// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importTemplate Tests With mixed imports 1`] = `"import TestDefault, { TestNamed_1, TestNamed_2 } from \\"./test\\""`;

exports[`importTemplate Tests With only default import 1`] = `"import TestDefault from \\"./test\\""`;

exports[`importTemplate Tests With only multiple named imports 1`] = `"import { TestNamed_1, TestNamed_2 } from \\"./test\\""`;

exports[`importTemplate Tests With only multiple unsorted named imports 1`] = `"import { TestNamed_1, TestNamed_2 } from \\"./test\\""`;

exports[`importTemplate Tests With only one named import 1`] = `"import { TestNamed } from \\"./test\\""`;
